torch/csrc/api/src/nn/modules/upsampling.cpp

```diff
@@ -0,0 +1,38 @@
+#include <torch/nn/modules/upsampling.h>
+
+#include <string>
+
+namespace F = torch::nn::functional;
+
+namespace torch {
+namespace nn {
+
+UpsampleImpl::UpsampleImpl(const UpsampleOptions& options_)
+    : options(options_) {}
+
+void UpsampleImpl::reset() {}
+
+void UpsampleImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::Upsample(";
+  if (!options.scale_factor().empty()) {
+    stream << "scale_factor=" << at::ArrayRef<double>(options.scale_factor());
+  } else {
+    stream << "size=" << at::ArrayRef<int64_t>(options.size());
+  }
+  stream << ", mode=" << c10::visit(enumtype::enum_name{}, options.mode()) << ")";
+}
+
+Tensor UpsampleImpl::forward(const Tensor& input) {
+  return F::interpolate(
+      input,
+      InterpolateOptions()
+          .size(options.size())
+          .scale_factor(options.scale_factor())
+          .mode(decltype(InterpolateOptions().mode())(options.mode()))
```

Comment:
curious does `.mode(options.mode())` work?

Reply:
The problem here is that `UpsampleOptions::mode_t` and `InterpolateOptions::mode_t` differ by one variant: `torch::kArea` (in Python` "area"`). According to the Python [`torch.nn.Upsample`](https://pytorch.org/docs/master/nn.html?highlight=upsample#torch.nn.Upsample) documentation, `"area"` is not an acceptable `mode` option, as `"area"` is commonly used as a decimation/downsampling mechanism. It happens to be passable because it is not checked in the module's `__init__`.

Would you like me to add `torch::kArea` as a variant in `Upsample::mode_t`? This would make your suggestion possible. :)

Reply:
Ah that makes sense, thanks for the catch! One thing that's a bit worrying to me is that I am not sure if the conversion from `UpsampleOptions::mode_t` to `InterpolateOptions::mode_t` is always well-defined :( Maybe we can create a temporary variable and use if-branching to set the value to be passed into `F::interpolate` explicitly :D

Reply:
Sure thing! I thought since `UpsampleOptions::mode_t` was a subset of `InterpolateOptions::mode_t` that this would be okay. I'll update shortly!

Reply:
Note: The `TORCH_ARG`-macros carry the `private:` access modifier over to the next line so that any following non-`TORCH_ARG` declaration gets "secretly" privatized. In this case, it hides
`typedef c10::variant<...> mode_t` (and similarly other option variants such as the `reduction_t`s) from users.

I had to place the `InterpolateOptions::mode_t` type alias declaration before any `TORCH_ARG` to expose the type publicly. This allowed me to create a temporary `InterpolateOptions::mode_t mode;` and ergonomically make the necessary update based on your suggestion.

Just thought I'd mention it because it may be useful to let users have access to these types: modes, reductions, etc. :D. Otherwise, `decltype` seems to be the only way to get this type information. Let me know what you think! :)

Reply:
Thanks so much for the catch! Yes I think we should establish the convention throughout the codebase that `typedef c10::variant<...> mode_t` should always happen in the first line of any `Options` that uses `c10::variant` (It is not enforced everywhere in the codebase right now, I opened a PR for `PadOptions` in https://github.com/pytorch/pytorch/pull/28760) :D